bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/bridge/VeluxBridgeActuators.java

```diff
@@ -0,0 +1,175 @@
+ */
+@NonNullByDefault
+public class VeluxBridgeActuators {
+    private final Logger logger = LoggerFactory.getLogger(VeluxBridgeActuators.class);
+
+    // Configuration constants
+
+    /**
+     * Limitation of Discovery on parts of the System table
+     *
+     * Whereas the parameter {@link org.openhab.binding.velux.things.VeluxKLFAPI#KLF_SYSTEMTABLE_MAX} represents the
+     * maximum size of the system table in general, for speed up of the discovery process
+     * a binding-internal limitation of number of possible devices is introduced.
+     */
+    private static final int VELUXBINDING_SYSTEMTABLE_MAX = 16;
+
+    // Type definitions, class-internal variables
+
+    /**
+     * Actuator information consisting of:
+     * <ul>
+     * <li>isRetrieved (boolean),
+     * <li>existingProducts ({@link VeluxExistingProducts}).
+     * </ul>
+     */
+    public class Channel { // NO_UCD (use private)
+        public VeluxExistingProducts existingProducts = new VeluxExistingProducts();
+    }
+
+    private Channel channel;
+
+    // Constructor methods
+
+    /**
+     * Constructor.
+     * <P>
+     * Initializes the internal data structure {@link #channel} of Velux actuators/products,
+     * which is publicly accessible via the method {@link #getChannel()}.
+     */
+    public VeluxBridgeActuators() {
+        logger.trace("VeluxBridgeActuators(constructor) called.");
+        channel = new Channel();
+        logger.trace("VeluxBridgeActuators(constructor) done.");
+    }
+
+    // Class access methods
+
+    /**
+     * Provide access to the internal structure of actuators/products.
+     *
+     * @return a channel describing the overall actuator situation.
+     */
+    public Channel getChannel() {
+        return channel;
+    }
+
+    /**
+     * Login into bridge, retrieve all products and logout from bridge based
+     * on a well-prepared environment of a {@link VeluxBridgeProvider}. The results
+     * are stored within {@link org.openhab.binding.velux.things.VeluxExistingProducts
+     * VeluxExistingProducts}.
+     *
+     * @param bridge Initialized Velux bridge (communication) handler.
+     * @return true if successful, and false otherwise.
+     */
+    @SuppressWarnings("null")
+    public boolean getProducts(VeluxBridge bridge) {
+        logger.trace("getProducts() called.");
+
+        GetProducts bcp = bridge.bridgeAPI().getProducts();
+        if (!bridge.bridgeInstance.veluxBridgeConfiguration().isBulkRetrievalEnabled || (bcp == null)) {
+            logger.trace("getProducts() working on step-by-step retrieval.");
+            GetProduct bcpSbS = bridge.bridgeAPI().getProduct();
+            for (int nodeId = 0; nodeId < VeluxKLFAPI.KLF_SYSTEMTABLE_MAX
+                    && nodeId < VELUXBINDING_SYSTEMTABLE_MAX; nodeId++) {
+                logger.trace("getProducts() {}.", new String(new char[80]).replace('\0', '*'));
+                try {
+                    Thread.sleep(2000);
```

Comment:
Thanks for your remarks about `Thread.sleep`. 
Am I right that your proposal is to create an own thread pool for device I/O instead of a `Thread.sleep() `within the openHab scheduled threads?

Reply:
You can also use the openHAB thread pool, but you better not block a thread but schedule a method to be executed two seconds later to execute the rest of the method. A thread / CPU can do a lot of other useful things in those two seconds.

Reply:
The sleeps are eliminated, now.

Reply:
Additionally I have checked that any sleeps only occur within a separate scheduler:

```
   /**
     * Dedicated thread pool for the long-running bridge communication threads.
     */
    private final ScheduledExecutorService handleScheduler = ThreadPoolManager
            .getScheduledPool(VeluxBindingConstants.BINDING_ID);


```